lib/vector-common/src/request_metadata.rs

```diff
@@ -94,5 +264,96 @@ impl<'a> Add<&'a RequestMetadata> for RequestMetadata {
 /// Objects implementing this trait have metadata that describes the request.
 pub trait MetaDescriptive {
     /// Returns the `RequestMetadata` associated with this object.
-    fn get_metadata(&self) -> RequestMetadata;
+    fn get_metadata(&self) -> &RequestMetadata;
+
+    /// Returns the owned `RequestMetadata`.
+    /// It is *very* important that this function is not called more than once for each
+    /// instance since it is likely that by taking the metadata it will replace the
+    /// current values with default metadata.
```

Comment:
🤔 could we add a protection to prevent calling that twice?

Independently of that, I realize this is a common trait that isn't explicitly tied to the sink service framework, but we could consider giving the explicit example of calling this function in the sink's service::call() function, since that seems to be the most likely use case.

Reply:
Given that this is a trait method, I'm not seeing a way one could do that? Each individual implementer could have it in an Option and then panic (or similar) if you take a `None`, but the trait can't enforce that..

Reply:
Potentially @bruceg suggestion to change this to `fn metadata_mut(&mut self) -> &mut RequestMetadata` would make more sense?

Reply:
> Given that this is a trait method

yeah, I realized that shortly after leaving the comment. Agree that enforcing that would be annoying.

> Potentially @bruceg suggestion to change this to fn metadata_mut(&mut self) -> &mut RequestMetadata would make more sense?

👍 👍 yesss

Reply:
Another possibility would be to store `Option<RequestMetadata>` in the types, and have the take method return the same. Then subsequent takes would return `None`.

Reply:
I've updated it to use `metadata_mut`.